hvl2wav/hvl2wav.1
.TH HVL2WAV "1" "September 2018" "HVL2WAV" "User Commands"
.SH NAME
HVL2WAV \- manual page for HVL2WAV
.SH SYNOPSIS
.B hvl2wav
[\fI\,options\/\fR] \fI\,<filename>\/\fR
.SH DESCRIPTION
HVL2WAV 1.2, replayer v1.8. Render a WAV audio file from HVL or AHX songs.
.SH OPTIONS
.HP
\fB\-ofile\fR Output to 'file'
.HP
\fB\-tm\fR:ss Timeout after m minutes and ss seconds
.IP
(overrides songend detection, defaults to 10:00
.TP
\fB\-fn\fR
Set frequency, where n is 8000 to 48000
.TP
\fB\-sn\fR
Use subsong n
.TP
\fB\-a\fR
Calculate optimal gain before converting
.TP
\fB\-xn\fR
Set stereo seperation (ahx only, no effect on HVL files)
.IP
0 = 0% (Mono)
1 = 25%
2 = 50%
3 = 75%
4 = 100% (Paula)


ChangeLog.txt
1.9 xx-xxx-xxxx
===============

- Added clamping to mixchunk() in the example replayer code (but not in the
  tracker proper as clipping noise is a good indicator to the composer to
  sort out the gain setting ;-)
- White noise is now much more accurate to the real AHX sound <passing_by>
- Example replayers come with accurate Amiga PAL/NTSC frequencies <passing_by>
- Filters now sound exactly like the real AHX ones <kode54 and passing_by>
- Fixed potential divide by zero <passing_by>
- Added workaround for Alt Gr handling on Windows. The windows SDL library
  sends dummy RCTRL key presses when you hit Alt Gr, so now right alt works
  as it should (thanks to AceMan)


1.8 07-May-2013
===============

- Fixed some warnings
- Fixed E1x and E2x fine slide commands (thanks to passing_by and kode54)
- Fixed broken stripping of 4xx upper nibble (thanks to passing_by). Due to
  a copy/paste bug, the fix for this in 1.6 only affected the left instrument
  column.
- Speed multiplier range is now 1-4 in the GUI. The HVL and AHX module formats
  only support this range anyway, so you couldn't save a module with a mult.
  of 5 or 6 (thanks to javamannen)
- Fixed some keyboard documentation bugs (thanks to javamannen)
- "Alt+Cursor Left" and "Alt+Cursor Right" now behave the same as AHX (suggested
  by javamannen)
- Updated all copyright dates and updated all URLs to hivelytracker.co.uk
- Removed all mentions of the sadly defunct IRIS homepage


1.7 27-Oct-2012
===============

All platforms:

- Fixed some crash bugs when closing tabs under certain circumstances
- Fixed some memory leaks
- Added an SDL wrapper that allows HivelyTracker to build for other
  platforms (Windows, OSX..)


OS4:

- Replaced some deprecated system calls with newer OS4 APIs


1.6 22-Oct-2008
===============

- Right since 1.0, EDx was broken because of how i'd implimented the
  two effect columns. Fixed. All replayers and plugins will need
  updating :-/ Thanks to skope for pointing this out.
- Fixed validity check for Cxx command, which broke in the ht replayer
  because the parameter is signed. (fixes the sidvibes ahx file, which
  used invalid Cxx commands). Thanks to megacz for pointing it out.
- The replayer now strips the upper nibble of 4xx instrument commands
  from "THX\0" AHX modules. These modules are saved from older versions
  of AHX that didn't have filters, and so if the upper nibble is set
  it activated filtering which is wrong since the module was created
  in an editor that didn't have filters. The AHX editor also does this
  stripping, but AFAIK WinAHX and its derivatives never did.


1.5 18-Aug-2008
===============

- A parameter in the undo system was declared as a signed byte instead
  of a signed word, which meant that some song parameters got corrupted
  when they went over 127. Editing songs over 127 positions long was
  buggered, for example. Can't believe i didn't spot this until now :(
- Added Protracker mod import. Just like the one in AHX, it simply strips
  out all PT commands that are not equivalent to AHX ones, and all
  instruments are set to a simple tone, so you'll still have a lot of
  work to do to make it sound nice.
- Added a size optimisation feature. It removes all unused instruments,
  and duplicate tracks. If you click with the right mousebutton, it'll
  additionally try and use the transpose column to remove tracks which
  are the same but transposed. However it doesn't know if sounds
  continue after the newly transposed track so you may have to manually
  fix any problems with that afterwards. Both modes keep the original
  version and open a new tab with the optimise version so you won't
  lose your work if it messes up. Both modes also use track 0 as the
  empty track as AHX and HVL both make smaller files if that is the
  case, however, if there are no empty tracks in the module, it won't
  keep track 0 blank unnecessarily.
- Added a new command! When "EF1" is placed on or after the last note
  in a track, the "transpose" value for that channel will be preserved
  until another note is played in the channel. This overrides the
  default AHX/HVL behaviour, which is to switch to the next transpose
  value as soon as the next position is reached.
- '=' now does row preview, like enter does in protracker
- '-' now toggles horizontal advance in the position editor
- Added "posedadvance", "notejump" and "inotejump" fields to the
  ht.prefs file. They don't show up in the prefs window, though, so
  you have to edit them manually (run hively 1.5 and quit it to make
  them show up in the file). "posedadvance" is either 0 or 1. When set
  to 0, horizontal advance is disabled by default.
  "defnotejump" is 0 to 9, and sets the default notejump in the pattern
  editor. "definotejump" is the same for the instrument editor.
  

1.4 24-Aug-2007
===============

- Playback no longer stops when the "Load Mod" filerequester is open
- Undo & Redo for almost all edit operations
- Cutting an area in the position editor cleared both columns of
  the left & right edges even if they weren't selected. Fixed.
- Cutting or pasting in the position editor didn't update the
  track editor even if the current tracks were changed by the
  operation. Fixed.
- If you change preference options that require the GUI to be closed
  and re-opened, it now does that instead of requiring you to quit
  and restart.
- Skin directory, initial song directory and initial instrument
  directory options all now can be selected via ASL directory requester.
- Right-clicking on a channel mute button now mutes all other channels
  except that channel. If it is already the only un-muted channel,
  all channels will be unmuted instead (suggested by Syphus).
- Added "9xx" panning command to the instrument performancelist. It
  only changes the panning position for that specific instrument. The
  next time an instrument is played in that channel, it returns to
  the last panning position set with the traditional "7xx" track

hvl2wav/hvl2wav.c
/*
** HVL2WAV
**
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "types.h"
#include "replay.h"

uint32 max_frames = 10 * 60 * 50;
uint32 framelen;
uint32 mix_freq = 44100;
uint32 subsong = 0, smplen = 0;
uint32 stereosep = 0;
BOOL autogain = FALSE;
BOOL use_songend = TRUE;
char *filename = NULL;
char *outname = NULL;
char wavname[1020];
char tmpname[1024];
int8 *mixbuf = NULL;
struct hvl_tune *ht = NULL;
FILE *of = NULL;
FILE *tf = NULL;

void write32( FILE *f, uint32 n )
{
#ifdef _BIG_ENDIAN_
  fprintf( f, "%c%c%c%c", (int)(n&0xff), (int)((n>>8)&0xff), (int)((n>>16)&0xff), (int)((n>>24)&0xff) );
#else
  fwrite( &n, 4, 1, f );
#endif
}

void write16( FILE *f, uint16 n )
{
#ifdef _BIG_ENDIAN_
  fprintf( f, "%c%c", n&0xff, (n>>8)&0xff );
#else
  fwrite( &n, 2, 1, f );
#endif
}

int main( int argc, char *argv[] )
{
  int i, j;
  int32 l;
  uint32 frm;

  for( i=1; i<argc; i++ )
  {
    if( argv[i][0] == '-' )
    {
      switch( argv[i][1] )
      {
        case 't':
          max_frames = atoi( &argv[i][2] ) * 50;
          j = 2;
          while( ( argv[i][j] >= '0' ) && ( argv[i][j] <= '9' ) ) j++;
          if( argv[i][j] == ':' )
          {
            max_frames *= 60;
            max_frames += (atoi( &argv[i][j+1] )*50);
          }

          if( max_frames == 0 )
          {
            printf( "Invalid timeout\n" );
            return 0;
          }

          use_songend = FALSE;
          break;
      
        case 'f':
          mix_freq = atoi( &argv[i][2] );
          if( ( mix_freq < 8000 ) || ( mix_freq > 48000 ) )
          {
            printf( "Frequency should be between 8000 and 48000\n" );
            return 0;
          }
          break;
        
        case 's':
          subsong = atoi( &argv[i][2] );
          break;
        
        case 'a':
          autogain = TRUE;
          break;
        
        case 'o':
          outname = &argv[i][2];
          if( outname[0] == 0 )
          {
            printf( "Invalid output name\n" );
            return 0;
          }
          break;
        
        case 'x':
          stereosep = atoi( &argv[i][2] );
          if( stereosep > 4 ) stereosep = 4;
          break;
      }
    } else {
      if( !filename ) filename = argv[i];
    }
  }

  if( !filename )
  {
    printf( "HVL2WAV 1.2, replayer v1.8\n" );
    printf( "By Xeron/IRIS (pete@petergordon.org.uk)\n" );
    printf( "\n" );
    printf( "Usage: hvl2wav [options] <filename>\n" );
    printf( "\n" );
    printf( "Options:\n" );
    printf( "  -ofile Output to 'file'\n" );
    printf( "  -tm:ss Timeout after m minutes and ss seconds\n" );
    printf( "         (overrides songend detection, defaults to 10:00\n" );
    printf( "  -fn    Set frequency, where n is 8000 to 48000\n" );
    printf( "  -sn    Use subsong n\n" );
    printf( "  -a     Calculate optimal gain before converting\n" );
    printf( "  -xn    Set stereo seperation (ahx only, no effect on HVL files)\n" );
    printf( "            0 = 0%% (Mono)\n" );
    printf( "            1 = 25%%\n" );
    printf( "            2 = 50%%\n" );
    printf( "            3 = 75%%\n" );
    printf( "            4 = 100%% (Paula)\n" );
    return 0;
  }

  framelen = (mix_freq*2*2)/50;
  mixbuf = malloc( framelen );
  if( !mixbuf )
  {
    printf( "Out of memory :-(\n" );
    return 0;
  }

  hvl_InitReplayer();

  ht = hvl_LoadTune( filename, mix_freq, stereosep );
  if( !ht )
  {
    free( mixbuf );
    printf( "Unable to open '%s'\n", filename );
    return 0;
  }

  if( !hvl_InitSubsong( ht, subsong ) )
  {
    hvl_FreeTune( ht );
    free( mixbuf );
    printf( "Unable to initialise subsong %ld\n", subsong );
    return 0;
  }

  if( autogain )
  {
    printf( "Calculating optimal gain...\n" );
    l = hvl_FindLoudest( ht, max_frames, use_songend );
    if( l > 0 )
      l = 3276700/l;
    else
      l = 100;
    
    if( l > 200 ) l = 200;
    ht->ht_mixgain = (l*256)/100;

    hvl_InitSubsong( ht, subsong );
    ht->ht_SongEndReached = 0;
  }

  printf( "Converting tune...\n" );

  if( outname )
  {
    strncpy( wavname, outname, 1020 );
  } else {
    if( ( strlen( filename ) > 4 ) &&
        ( ( strncasecmp( filename, "hvl.", 4 ) == 0 ) ||
          ( strncasecmp( filename, "ahx.", 4 ) == 0 ) ) )
      strncpy( wavname, &filename[4], 1020 );
    else
      strncpy( wavname, filename, 1020 );

    if( ( strlen( wavname ) > 4 ) &&
        ( ( strcasecmp( &wavname[strlen(wavname)-4], ".hvl" ) == 0 ) ||
          ( strcasecmp( &wavname[strlen(wavname)-4], ".ahx" ) == 0 ) ) )
      wavname[strlen(wavname)-4] = 0;
    strcat( wavname, ".wav" );
  }

  strcpy( tmpname, wavname );
  strcat( tmpname, ".tmp" );

  tf = fopen( tmpname, "wb" );
  if( !tf )
  {
    hvl_FreeTune( ht );
    free( mixbuf );
    printf( "Unable to open '%s' for output\n", tmpname );
    return 0;
  }

  frm = 0;
  smplen = 0;
  while( frm < max_frames )
  {
    if( ( use_songend ) && ( ht->ht_SongEndReached ) )
      break;

    hvl_DecodeFrame( ht, mixbuf, &mixbuf[2], 4 );

#ifdef _BIG_ENDIAN_
    // The replayer generates audio samples in the same
    // endian mode as the CPU. So on big endian systems
    // we have to swap it all to stupid endian for the
    // WAV format.
    for( i=0; i<framelen; i+=2 )
    {
      int8 k;
      k = mixbuf[i]; mixbuf[i] = mixbuf[i+1]; mixbuf[i+1] = k;
    }
#endif

    fwrite( mixbuf, framelen, 1, tf );
    smplen += framelen;
    frm++;
  }

  fclose( tf );

  of = fopen( wavname, "wb" );
  if( !of )
  {
    hvl_FreeTune( ht );
    free( mixbuf );
    unlink( tmpname );
    printf( "Unable to open '%s' for output\n", wavname );
    return 0;
  }

  tf = fopen( tmpname, "rb" );
  if( !tf )
  {
    fclose( of );
    unlink( wavname );
    unlink( tmpname );
    hvl_FreeTune( ht );
    free( mixbuf );
    printf( "Unable to open '%s' for input\n", tmpname );
    return 0;
  }

  fprintf( of, "RIFF" );
  write32( of, smplen+36 );
  fprintf( of, "WAVE" );
  fprintf( of, "fmt " );
  write32( of, 16 );
  write16( of, 1 );
  write16( of, 2 );
  write32( of, mix_freq );
  write32( of, mix_freq*4 );
  write16( of, 4 );
  write16( of, 16 );
  fprintf( of, "data" );
  write32( of, smplen );

  for( i=0; i<smplen; i+=framelen )
  {
    fread( mixbuf, framelen, 1, tf );
    fwrite( mixbuf, framelen, 1, of );
  }

  fclose( of );
  fclose( tf );
  unlink( tmpname );
  hvl_FreeTune( ht );
  free( mixbuf );
  printf( "ALL DONE!\n" );
  return 0;
}
